Share / link button

wbazant opened the pull request:
Closes #636
Closes #517

Big step towards #246!

- UI for mobile and desktop with a link icon and 'copy to clipboard' box
- preserve parameters in settings and filter into mostly readable format, e.g. ?map=satellite
- a long encoder/decoder that converts selections of types into strings like 'default', 'all', 'all-123.456', 'h136'
- also support legacy param names

ezwelty commented:
@wbazant Very exciting ⚡ 

## Interface

Seems fine to me. Small display nitpick is that the copy button should have the same height as the URL field and the right border of the field looks strange:

<img width="159" alt="image" src="https://github.com/user-attachments/assets/9e1b4a90-87e7-4722-9555-7b6c88b6b0af" />

I can't help wonder why we wouldn't (also) immediately update the URL in the browser bar for many of these, but that's another discussion.

## New URL params

* [ ] overlay (transit): Selecting bicycle crashes the app
* [x] map (terrain, hybrid, osm-standard, osm-toner-lite)
* [x] latitude, longitude, zoom
* [x] labels=true
* [x] poi=true
* [x] muni=false
* [x] /locations/id
* [x] /settings

As a general question, does it make sense for defaults to go unmentioned in the URL? They are shorter but less stable/explicit (e.g. default map changes to OSM and user expected Google Map).

## New types param

* [ ] Why period and not comma or pipe as used elsewhere (legacy website, API)?
* [ ] lower-level parent and all children: http://localhost:3000/?types=1052.3478.4456.3767.3627 – Why is it just a list of types in this case?
* [ ] single category (?c=freegan): http://localhost:3000/map/@40.1125785,-88.2287926,4z?types=h1485.h1628.5284.7683.1479.660.3820.6190.7456.2.836.5389.3165.5090.821.8480.2085.1385.852.853.846.5033.839.843.8532.4265.7734.4678.845.837.838.8887.5796.8892.7687.8500.7686.4247.7684.3151.7302.7444.8894.8889.3258.8893.7441.7737.7680.6129.7585.7735.8890.8888.6117.5607.8886.7574.8485.8685.7462.8891.6364.7573.7835.6213.5356.7733 – Is it intentional that the categories are not used? This is very different to ?c=freegan
* [x] default (also ?c=freegan,forager): http://localhost:3000 (but see above about defaults)
* [x] all: http://localhost:3000/?types=all
* [x] none: http://localhost:3000/?types=
* [x] one: http://localhost:3000/?types=1
* [x] two: http://localhost:3000/?types=1.2
* [x] top-level parent and all children: http://localhost:3000/?types=h843
* [x] default minus one: http://localhost:3000?types=default-2
* [x] default minus one top-level parent and all children: http://localhost:3000?types=default-h2027

Do you think much more could be gained from ranges (e.g. 1-3 means 1, 2, and 3)? I ask because - is here used as a negation. But another symbol could be used instead (e.g. 1:3 or 1_3).

## #517 Legacy map params

* [ ] http://localhost:3000/?z=15&y=47.41369&x=8.56009 → http://localhost:3000/map/@8.5600900,47.4136900,15z: Bug! Params `x` and `y` are longitude and latitude, so should map to `@{y},{x},{z}z`
* [ ] Combinations of 1–2 of `x`, `y` and `z` are ignored although legacy website overrides defaults granularly. Examples:
  * http://localhost:3000/?y=47.41369&x=8.56009
  * http://localhost:3000/?z=15
* [x] http://localhost:3000/?m=false
* [x] http://localhost:3000/?f=1
* [x] http://localhost:3000/?f=1,2
* [x] http://localhost:3000/?c=forager
* [x] http://localhost:3000/?c=freegan
* [x] http://localhost:3000/?c=grafter
* [x] http://localhost:3000/?c=honeybee
* [x] http://localhost:3000/?c=forager,freegan
* [x] http://localhost:3000/?c=freegan&f=1 (`f` overrides `c`)
* [x] http://localhost:3000/?t=roadmap
* [x] http://localhost:3000/?t=hybrid
* [x] http://localhost:3000/?t=terrain
* [x] http://localhost:3000/?t=toner-lite
* [x] http://localhost:3000/?t=osm
* [x] http://localhost:3000/?l=true

ezwelty commented:
Later I had an idea that the type selection could be expressed in terms of user interaction. Something like `{preset}` followed by a series of `[-+]` and either `{type.id}`, `h{type.id}`, `{preset}` (if added or removed from selection, rather than replacing current selection), or `q'{search}'` (if all results of search were added or removed from selection). But then I realized that the latter would be locale (and search-algorithm) specific and that this strategy completely breaks down with e.g. (de)selecting all types in current map bounds.

ezwelty commented:
p.s. I found myself really wishing for an "Only selected" option in the type tree when testing URLs with only a few types selected.

wbazant commented:
Thanks for the thorough testing!
> copy button should have the same height as the URL field and the right border of the field looks strange

fixed both - padding inside input was messing up the borders, also I made input have the same height as the button. Thanks!

>  overlay (transit): Selecting bicycle crashes the app

fixed typo 'BicycleLayer' -> 'BicyclingLayer', thanks!

>  Why period and not comma or pipe as used elsewhere (legacy website, API)?

comma in params gets encoded, making the URLs longer and messier. Dot doesn't and makes this part of the URL more readable - still cryptic but not as overwhelming

> lower-level parent and all children: http://localhost:3000/?types=1052.3478.4456.3767.3627 – Why is it just a list of types in this case?

I've originally not done it to keep implementation simpler (making sure that selecting a grandparent doesn't also include parents in the URL) but I've described the problem to AI and it gave me a solution, so that's fixed!

> Is it intentional that the categories are not used?

It was intentional, I thought that since there's no way to select a category in the UI, it's unlikely the encoder will choose that as the shortest representation.

> Do you think much more could be gained from ranges (e.g. 1-3 means 1, 2, and 3)?

I think no - there's no way to select them like that, so I don't expect big ranges will come up in URLs people end up constructing.

>  Bug! Params x and y are longitude and latitude, so should map to @{y},{x},{z}z

fixed, thanks for spotting!


> Combinations of 1–2 of x, y and z are ignored although legacy website overrides defaults granularly

changed behaviour to match live site

ezwelty commented:
@wbazant This is looking great, and my only remaining suggestions touch on the encoder/decoder.

> comma in params gets encoded, making the URLs longer and messier. Dot doesn't and makes this part of the URL more readable - still cryptic but not as overwhelming

Actually `,` has no special meaning in the query component of a URL and thus does not need to be encoded. (sources: [STD 66](https://tools.ietf.org/html/std66), https://stackoverflow.com/a/45702126, ...). Because of your use of `-` as remove from selection, you could consider replacing `.` with `+`.

I notice that the encoder doesn't use the `h{id}` notation when the reference selection includes some but not all of the hierarchy. Since the end result is known and explicit (hierarchy either all full or all empty), the reference state does not matter. Is it possible to use `h{id}` also in this case? I believe this change could make a big difference.

Here is the current outcome of default + all Maple (http://localhost:3000/map?types=default.h175):

http://localhost:3000/map?types=default.h2268.1026.1052.2655.1838.3439.3102.1062.1056.3241.1050.2656.1013.3673.3498.3776.1157.1100.1059.3677.3844.917.8730.7386.8725.7347.3530.3492.3537.2021.2269.2270

wbazant commented:
Let's go with dot then, I'm not sure why , and + get encoded:
![Screenshot from 2025-04-09 16-22-34](https://github.com/user-attachments/assets/f61b5e90-95b8-4448-9e13-6ecb726a4cf4)

I added non-top hierarchies and hierarchies that overlap with default, for adding and for removing. So now adding all of a tree that's partially selected by default, or removing all, should get shortened.

ezwelty commented:
@wbazant Funny, guess the the Javascript URL class implementation is overly conservative. All browsers I know leave `+` and `,` in the params in place.

Thanks for tweaking the encoder further! I'm gonna stop pestering you now and merge. Nice work!
